physical/mongo.go

```diff
@@ -0,0 +1,1121 @@
+		c := b.session.DB(b.Database).C(b.Collection)
+		err = c.EnsureIndex(keyIndex)
+		if err != nil {
+			b.logger.Error(fmt.Sprintf("physical/mongo: could not EnsureIndex() for 'key': %v", err))
+			return nil, err
+		}
+
+		err = c.EnsureIndex(lastCheckedInIndex)
+		if err != nil {
+			b.logger.Error(fmt.Sprintf("physical/mongo: could not EnsureIndex() for 'lastCheckedIn': %v", err))
+			return nil, err
+		}
+		b.ensuredDataIndices = true
+	}
+
+	return b.session, nil
+}
+
+// this part was derived from and adjusted for our needs from:
+//   builtin/logical/mongodb/backend.go
+//
+func (b *MongoBackend) activeSessionMsgReceiver() (*mgo.Session, error) {
+	b.sessionMsgReceiverLock.Lock()
+	defer b.sessionMsgReceiverLock.Unlock()
+
+	if b.sessionMsgReceiver != nil {
+		if err := b.sessionMsgReceiver.Ping(); err == nil {
+			return b.sessionMsgReceiver, nil
+		}
+		b.sessionMsgReceiver.Close()
+		b.ensuredMsgCapped = false
+	}
+
+	// we need to establish a new session
+	b.logger.Info("physical/mongo: establishing new MongoDB session for Broadcast Message Receiver")
+	dialInfo, err := b.parseMongoURI(b.Url)
+	if err != nil {
+		b.logger.Error(fmt.Sprintf("physical/mongo: could not parse MongoDB URI: %v", err))
+		return nil, err
+	}
+
+	b.sessionMsgReceiver, err = mgo.DialWithInfo(dialInfo)
+	if err != nil {
+		b.logger.Error(fmt.Sprintf("physical/mongo: could not establish connection to MongoDB: %v", err))
+		return nil, err
+	}
+	// TODO: adjust these values
+	b.sessionMsgReceiver.SetSyncTimeout(1 * time.Minute)
+	b.sessionMsgReceiver.SetSocketTimeout(1 * time.Minute)
+
+	if b.haEnabled && !b.ensuredMsgCapped {
+		b.logger.Info(fmt.Sprintf("physical/mongo: ensuring MongoDB collection '%v' exists and is capped", b.CollectionMsg))
+		// check if collection exists, create if not
+		// if it exitst, check if it is a capped collection, if not use convertToCapped (?)
+		db := b.sessionMsgReceiver.DB(b.Database)
+		colls, err := db.CollectionNames()
+		if err != nil {
+			b.logger.Error(fmt.Sprintf("physical/mongo: could not retrieve list of collection names: %v", err))
+			return nil, err
+		}
+
+		exists := false
+		for _, coll := range colls {
+			if coll == b.CollectionMsg {
+				exists = true
+				break
+			}
+		}
+
+		if exists {
+			b.logger.Info(fmt.Sprintf("physical/mongo: MongoDB collection '%v' exists, ensuring it is capped", b.CollectionMsg))
+			var result CappedResult
+			err = db.Run(bson.D{bson.DocElem{Name: "collStats", Value: b.CollectionMsg}}, &result)
+			if err != nil {
+				b.logger.Error(fmt.Sprintf("physical/mongo: could not retrieve collStats: %v", err))
+				return nil, err
+			}
+
+			if !result.Capped {
+				b.logger.Info(fmt.Sprintf("physical/mongo: MongoDB collection '%v' is not capped, running convertToCapped", b.CollectionMsg))
+				var result2 bson.M
+				err = db.Run(bson.D{bson.DocElem{Name: "convertToCapped", Value: b.CollectionMsg}, bson.DocElem{Name: "size", Value: 4096}}, &result2)
```

Comment:
This collection is independent from the locking mechanism. The purpose of this collection is to be a broadcast channel for state changes on the lock / leader (kind of like pub/sub). It's basically ensuring that when a leader is stopped gracefully or there are other reasons for a leader to step down that they can send out a message. That said every follower is also watching the lock and in case the leader died (= did not update the lock in a certain amount of time), they will also use this channel to broadcast to start the "leader election" (in quotes here as this is not the same as leader election in raft e.g.).

Now to the problem:
- you want to keep this collection small because otherwise it's just growing your database with useless information
- you can specify a max bytes during conversion and creating
- and if I understand the mongo docs correctly, you can only specify max docs on creation

Unfortunately we need to make a trade-off here. The other possibility is to fail the startup if the collection is not capped and was not created by us before.

That said, I will double-check again that my understanding of mongo here is correct.